core/src/main/scala/kafka/server/ReplicaManager.scala

```diff
@@ -538,13 +541,21 @@ class ReplicaManager(val config: KafkaConfig,
           val initialLogEndOffset = localReplica.logEndOffset
           val logReadInfo = localReplica.log match {
             case Some(log) =>
-              val adjustedFetchSize = if (Topic.isInternal(topic) && !readOnlyCommitted) Math.max(fetchSize, log.config.maxMessageSize) else fetchSize
+              val adjustedFetchSize =
```

Comment:
adjustedFetchSize probably should be the smaller of fetchSize and the remaining bytes in fetchMaxBytes.